### PR TITLE
MDEV-341: 64-bit suport in FROM_UNIXTIME()

### PR DESCRIPTION
**Pull request by ghost**
- [x] *The Jira issue number for this PR is: MDEV-341*

## Description
This patch is a somewhat crude but successful attempt to add 64-bit support for the functions FROM_UNIXTIME() and UNIX_TIMESTAMP(). This rather old issue is becoming more and more relevant as the deadline of 2038 approaches, at which point these two functions cease to function properly.

## How can this PR be tested?
Executing `select unix_timestamp('2050-01-01 12:00:00');` should return the value "2524651200".
Executing `select from_unixtime(2524651200);` should return the value "2050-01-01 12:00:00".

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*
- [x] *This bug has existed since 1995. Basing this PR on 29e300ebbfd60 does not make much sense.*

## Backward compatibility
I made sure not to affect the storage format of column types TIMESTAMP and DATETIME, something that is of pretty big concern. Only the output of the two functions FROM_UNIXTIME() and UNIX_TIMESTAMP() have been changed, both functions will no longer return NULL for values exceeding INT32_MAX (2038-ish).

Users that rely on the fact that these functions return NULL for values exceeding INT32_MAX will run into problems, as one might expect.

### CODE REVIEW

**Review by dlenski — CHANGES_REQUESTED** [review]
> # How can this PR be tested?
> 
> Executing `select unix_timestamp('2050-01-01 12:00:00');` should return the value "2524651200".
> Executing `select from_unixtime(2524651200);` should return the value "2050-01-01 12:00:00".

Could you add these to the automated mysql-test suite?

It should be very straightforward to add a few lines to the existing tests of time-related functions:

- Tests: https://github.com/MariaDB/server/blob/HEAD/mysql-test/main/func_time.test
- Expected results: https://github.com/MariaDB/server/blob/HEAD/mysql-test/main/func_time.result

File: sql/item_timefunc.cc
Comment (FooBarrior) [inline sql/item_timefunc.cc:1215]: I dont think a change like this is a good idea.
I may see two points:
1. there may be some storage format tied to `my_time_t` size;
2. some other conversion issues

Other than that I don't see why not just redefine `my_time_t` to something larger.

Anyways, neither I do think that we'll not move to 128-bit architecture to 2038.
At least, making a 128-bit cache line is rather a demand.
With this I wouldn't think that `long long` will remain 64 bit, especially with guarantee. Most likely, it will depend on the implementation, just like now.

So at least you should use something constantly-sized, if you want a strictly 64-bit large output.
`uint64_t` is good enough (I think we support `<stdint.h>`).

I would also think of making a dedicated type like `my_time_t_64` or something, to which we will eventlually move the rest of the code base.

**Comment by dbuteyn** [comment]
> 
> Could you add these to the automated mysql-test suite?
> 
> It should be very straightforward to add a few lines to the existing tests of time-related functions:

I can, still going through all the failed tests however. A lot of them expect NULL for dates past 2038 and will need adjustment.

File: sql/item_timefunc.cc
Comment (dbuteyn) [reply sql/item_timefunc.cc:1215]: > I dont think a change like this is a good idea. I may see two points:
> 
>     1. there may be some storage format tied to `my_time_t` size;
That function is only called from within Item_func_unix_tijmestamp itself, it has no relation to the storage format.

>     2. some other conversion issues
I am still working on verifying and correcting all tests. Can you be more specific as to which conversion issues you are expecting?

> Other than that I don't see why not just redefine `my_time_t` to something larger.

Redefining my_time_t to something else has very large implications, notably it means a header change for 3rd-party software. But also tons of other MySQL internals will be affected (including the storage format) which I don't think is a good idea.
 
> I would also think of making a dedicated type like `my_time_t_64` or something, to which we would eventlually move the rest of the code base.

Not a bad idea, though if you look carefully you'll see that all I really did was replace the internal conversion done by UNIX_TIMESTAMP() from TIMESTAMP to DATETIME and enable FROM_UNIXTIME() parsing of dates > 2038. I tried hard to make sure nothing else would be affected, especially the storage format of TIMESTAMP. DATETIME columns themselves are fine, they already support dates from 0000 to 9999 if I recall correctly.

**Comment by dbuteyn** [comment]
While trying to figure out what to do with negative timestamps (technically legal since time_t, timeval and friends are signed), I happen to notice that MySQL 8.0.28 added 64-bit support for UNIX_TIMESTAMP() and FROM_UNIXTIME():

https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_from-unixtime

> On 32-bit platforms, the maximum useful value for unix_timestamp is 2147483647.999999, which returns '2038-01-19 03:14:07.999999' UTC. On 64-bit platforms running MySQL 8.0.28 or later, the effective maximum is 32536771199.999999, which returns '3001-01-18 23:59:59.999999' UTC. Regardless of platform or version, a greater value for unix_timestamp than the effective maximum returns 0.

https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_unix-timestamp

> Prior to MySQL 8.0.28, the valid range of argument values is the same as for the TIMESTAMP data type: '1970-01-01 00:00:01.000000' UTC to '2038-01-19 03:14:07.999999' UTC. This is also the case in MySQL 8.0.28 and later for 32-bit platforms. For MySQL 8.0.28 and later running on 64-bit platforms, the valid range of argument values for UNIX_TIMESTAMP() is '1970-01-01 00:00:01.000000' UTC to '3001-01-19 03:14:07.999999' UTC (corresponding to 32536771199.999999 seconds). 

I'm not sure whether I should continue to invest time on this issue and whether it is legal to port code from MySQL to MariaDB.

**Comment by vuvova** [comment]
Yes, it's legal. MySQL is GPLv2, MariaDB is GPLv2, you can take move GPLv2 code from one project to the other.

Code base has diverged significantly, though. I _might_ be easier to finish your implementation, instead of porting from MySQL

**Comment by LinuxJedi** [comment]
This might need some collaboration with @montywi as he is currently working on expansion of the TIMESTAMP column and therefore may well touch related functions (I'll check with him shortly). But I suspect it might be easier to make this an unsigned instead of a signed, given that negatives aren't supported anyway.

**Comment by dbuteyn** [comment]
@LinuxJedi : MariaDB docs stat that the `DATETIME` type supports dates starting from 1000-01-01. Unix timestamp -386380800 is supposedly 04-10-1957 which is not an unreasonable date for someone working with historical data and would be a valid value for a DATETIME column. Users would expect `insert into table_with datetime values (from_unixtime(-386380800))` to work.

I would leave the `TIMESTAMP` type itself alone, it is fine as-is. If you really want to support timestamps > INT_MAX , adding a `TIMESTAMP64` type probably is the safer option in terms of compatibility.

**Comment by LinuxJedi** [comment]
@dbuteyn I've discussed this with Monty today, and he agrees with you that an even wider range is needed. He suggested waiting until MDEV-32188 has merged (which will extend it a bit, but not far enough). Then we can work on getting your work ready on top of this. He says he expects the review and merge to be done in about a month due to the upcoming MariaDB ServerFest / Unconference.

I hope this helps. Please let use know if you have any questions.

**Comment by dbuteyn** [comment]
@LinuxJedi Thank you for the update, appreciated.

File: sql/item_timefunc.cc
Comment (dr-m) [reply sql/item_timefunc.cc:1215]: A 128-bit (16-byte) cache line would be extraordinarily small, I think. IA-32 and AMD64 implementations typically have 64-byte (512-bit) cache lines. On other ISAs such as ARMv8 or s390x, the cache line can be even larger.

Comment (dbuteyn) [reply sql/item_timefunc.cc:1215]: I don't see how cache lines are related to a function argument?

Comment (dr-m) [reply sql/item_timefunc.cc:1215]: Maybe @FooBarrior was thinking about the alignment of the pointer target? Based on my experience from [MDEV-28091](https://jira.mariadb.org/browse/MDEV-28091), normal `malloc()` might already return 128-bit aligned memory on 64-bit systems. I believe that 64-bit alignment is guaranteed already on 32-bit systems.

Comment (dbuteyn) [reply sql/item_timefunc.cc:1215]: The pointer argument is the return value for this function. The function itself returns true/false indicating success/failure. Since the caller asked for the timestamp value, this value has to be returned somehow.

@FooBarrior was referring to `longlong` being possibly larger than 64-bits on 128-bit architectures and suggested using `(u)int64_t` instead to define the exact size for `seconds`. This suggestion is valid and correct.

**Pull request closed without merging**
